pkg/kubelet/pod_workers_test.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"sync"
+	"testing"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/dockertools"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/types"
+)
+
+func newPod(uid, name string) api.BoundPod {
+	return api.BoundPod{
+		ObjectMeta: api.ObjectMeta{
+			UID:  types.UID(uid),
+			Name: name,
+		},
+	}
+}
+
+func TestUpdatePod(t *testing.T) {
+	fakeDocker := &dockertools.FakeDockerClient{}
+	fakeDockerCache := dockertools.NewFakeDockerCache(fakeDocker)
+	waitGroup := sync.WaitGroup{}
+
+	lock := sync.Mutex{}
+	processed := make(map[types.UID][]string)
+
+	podWorkers := newPodWorkers(
+		fakeDockerCache,
+		func(pod *api.BoundPod, containers dockertools.DockerContainers) error {
+			func() {
+				lock.Lock()
+				defer lock.Unlock()
+				processed[pod.UID] = append(processed[pod.UID], pod.Name)
+			}()
+			waitGroup.Done()
+			return nil
+		})
+
+	// Check whether all pod updates will be processed.
+	numPods := 20
+	for i := 0; i < numPods; i++ {
+		for j := i; j < numPods; j++ {
+			waitGroup.Add(1)
+			podWorkers.UpdatePod(newPod(string(j), string(i)))
+		}
+	}
+	waitGroup.Wait()
+
+	if len(processed) != 20 {
+		t.Errorf("Not all pods processed: %v", len(processed))
+		return
+	}
+	for i := 0; i < numPods; i++ {
+		uid := types.UID(i)
+		if len(processed[uid]) != i+1 {
+			t.Errorf("Pod %v processed %v times", i, len(processed[uid]))
+			continue
+		}
+		for j := 0; j <= i; j++ {
+			if processed[uid][j] != string(j) {
+				t.Errorf("Pod %v: incorrect order %v, %v", i, j, processed[uid][j])
+			}
+		}
+	}
+}
+
+func TestForgetNonExistingPodWorkers(t *testing.T) {
+	fakeDocker := &dockertools.FakeDockerClient{}
```

Comment:
Can we use a helper for this? It is very similar to the one above and we may be able to just use the same one.